### ISSUE TITLE
Refactor/improve npmrc handling

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#9145]
**What would you like Renovate to be able to do?**

Parse `.npmrc` during npm extract phase and not use it at all in npmjs datasource.

**Did you already have any implementation ideas?**

npm extract should detect all relevant `.npmrc` config and return it as part of extract, e.g.
- For scope instructions, use that to determine the `registryUrl` of affected deps
- Use hostRules and/or packageRules (TBD)

npm datasource should not need to read from the file system, meaning:
- Either extracted dependencies embed their scoped registryUrl, or a package rule is used to determine the registryUrls
- Any authentication in the `.npmrc` is already configured elsewhere (e.g. hostRules)

We need to consider following situations:
- No `.npmrc` in registry, but scope and/or authentication are required in renovate config
- `.npmrc` in registry contains scope info but authentication is missing
- `.npmrc` in registry uses env substituation for authentication
- Any

I would also like to deprecate/migrate the use of `npmrc` and `npmToken` fields - to be replaced by something more generic.

**Comment by rarkins** [comment]
Here is where the datasource determines which `registryUrl` should be used: https://github.com/renovatebot/renovate/blob/6ebe8ba508d5fc2dc46e7592a014f7e0bbdea913/lib/datasource/npm/get.ts#L83-L88

This should instead be determined at extract time.

Here is where the datasource determines auth from the `.npmrc`: https://github.com/renovatebot/renovate/blob/6ebe8ba508d5fc2dc46e7592a014f7e0bbdea913/lib/datasource/npm/get.ts#L115-L120

Could these be set via `hostRules`, and should that be done during extract?

Should we parse `.npmrc` manually, or is there an acceptable npm lib to do it?

**Comment by rarkins** [comment]
@viceice what are your thoughts on deprecating `npmrc` as a config field? Options:
- Document that if scoped lookups are required then these should be committed to the repo in a `.npmrc` (doesn't mean auth needs to be committed.. just scope instructions)
- Allow converting of `packageRules` to `.npmrc` scope instructions via some type of convention. e.g. packageRule follows convention:

```json
{
  "hostType": "npm",
  "packagePatterns": "^@somescope\/",
  "registryUrls": ["https://my.custom.registry/"]
}
```

converts to:

```
@somescope:registry=https://my.custom.registry/
```

**Comment by viceice** [comment]
I think scopes should be in repo `.npmrc`, also the default registry should be read from `.npmrc` if configured (extraction phase)

I'm ok to fully ignore repo `.npmrc` and force users to configure packageRules, so no npmrc handling needed here.

On artifacts: write all npm hostRules to parent of repo / temp dir and allow usage like dotnet manager. If this not doable, write it to repo npmrc as currently done.

**Comment by rarkins** [comment]
> I think scopes should be in repo `.npmrc`, also the default registry should be read from `.npmrc` if configured (extraction phase)

I expect there would be *some* users who have no `.npmrc` in their repos and have default registry and/scope configured in individual developer machines. So in such a case it wouldn't work until they add the non-token parts of the `.npmrc` to the repo.

> I'm ok to fully ignore repo `.npmrc` and force users to configure packageRules, so no npmrc handling needed here.

I hadn't thought of being so strict, but maybe it's a good idea. The following scenarios which "just work" today would then not?
- scoped registry line in .npmrc is present, but doesn't require authentication - in which case we'll now incorrectly look up the default registry by default (unlikely?)
- a token is configured in the repo's `.npmrc` (bad practice..)

> On artifacts: write all npm hostRules to parent of repo / temp dir and allow usage like dotnet manager. If this not doable, write it to repo npmrc as currently done.

IIRC npm doesn't do a simple "find up" for `.npmrc`. I think it instead looks in order for:
- present working directory (/path/to/my/project/.npmrc)
- per-user configuration file (~/.npmrc)
- global configuration file (/etc/npmrc)

Also I think we might need to *override* any `.npmrc` in the repo, which would only work if we literally override it in the file system.

**Comment by rarkins** [comment]
Something we could do as a transition stage:
- Add the extra intelligence to the extra phase (including registryUrls, hostRules, etc)
- Keep the existing logic in the datasource
- Compare the existing logic result for registryUrl and auth to what we get from the "new" approach and log if they differ

I could suppress the logs in the app for users but capture it myself to see if we can solve any of the problems it finds.

**Comment by rarkins** [comment]
Here's where we use `NPM_TOKEN` today:

[If the decrypted config includes `npmToken`, and there's a *configured* `.npmrc` with `${NPM_TOKEN}`, then we substitute it in the string](https://github.com/renovatebot/renovate/blob/74d7691befa635c1220d8569936273bfb950e748/lib/config/decrypt.ts#L51-L58). Applicable only to the app.

[If there's no authentication from `.npmrc` parsing in the npm datasource, we apply `NPM_TOKEN` authorization no matter what the host is](https://github.com/renovatebot/renovate/blob/7d6870a6d06f0797520fb339120c7aaa4a1d3407/lib/datasource/npm/get.ts#L121-L127). Applicable only to self-hosted. The alternative here is to configure a host rule for the host or npm hostType.

We pass `NPM_TOKEN` through to child processes when generating lock file updates using [npm](https://github.com/renovatebot/renovate/blob/40a59cc9e0687a6302a5a5992949564af6b507c7/lib/manager/npm/post-update/npm.ts#L77), [pnpm](https://github.com/renovatebot/renovate/blob/40a59cc9e0687a6302a5a5992949564af6b507c7/lib/manager/npm/post-update/pnpm.ts#L56), [yarn](https://github.com/renovatebot/renovate/blob/40a59cc9e0687a6302a5a5992949564af6b507c7/lib/manager/npm/post-update/yarn.ts#L126) and [lerna](https://github.com/renovatebot/renovate/blob/40a59cc9e0687a6302a5a5992949564af6b507c7/lib/manager/npm/post-update/lerna.ts#L102). Applicable only to self-hosted.

The alternative here is less clear. If we don't pass the `NPM_TOKEN` then it would "break" anyone which is using variable interpolation to substitute `${NPM_TOKEN}` in the `.npmrc` or perhaps other files. The only alternative in such cases is either:
- Read the `.npmrc`, replace `${NPM_TOKEN}` (somehow, using a hostRules value), and then temporarily replace it, or
- Overwrite the `.npmrc` with something else, but presumably that something else would still need a token in it

In other words, we can't improve `NPM_TOKEN` handling independently of `.npmrc` refactoring

**Comment by rarkins** [comment]
Proposal: 
- the npm manager finds any `.npmrc` file next to `package.json` during extract phase and saves it into `managerData.npmrc`
- the npm datasource does not use `.npmrc` in any way - file system or config
- the npm manager reads the `.npmrc` file back during artifacts updating, massages it if necessary, and writes it back

Question is whether or how to handle a *configured* `npmrc`. For maximum backwards compatible or as a stepping stone we'd keep supporting it, with the logic being that we write any configured `npmrc` over the top of what we detected with `managerData.npmrc`.

The next questions are: how would the datasource know (a) which registryUrl, and (b) which authorization to use?

For registryUrl, it seems we'd need to parse the `.npmrc` during extract and either:
- return `registryUrls` as a result of the extract phase, or
- return a package rule bound to the path of this package.json

The first approach seems cleaner.

For authentication, the npm extract function could set hostRules if it finds any actual credentials in the `.npmrc` (which wouldn't be best practice to commit to git..). The only problem here is that in theory in a monorepo different packages could have different authentication for the same host, but that's maybe an acceptable limitation considering how rare that should be?

Final stage is the writing of `.npmrc`. For each `package.json`, we'd check if there's a `config.npmrc` and use it, otherwise a `managerData.npmrc`, and then we'd massage the content: 
- Look up any packageRules which have `matchDatasources=npm`, and `matchPackagePatterns` starting with `^@`, no other match rules, and `registryUrls` of length 1. If found, massage `registry=` lines into the starting `.npmrc` content
- Look up any hostRules where `hostType=npm` and add or replace `auth` lines

Note: in any case we'd deprecate the use of "home directory" or "global" `.npmrc` files, in a major release.

**Comment by rarkins** [comment]
Other approaches:

### Backwards-compatible `.npmrc` handling

In this scenario, we'd stop using the external library which reads from the file system during datasource lookup and instead find the home directory `~/.npmrc` ourselves the first time it's needed.

### Keep supporting a "global" `.npmrc`, but it must be in config

In this scenario, `~/.npmrc` is never read by Renovate or its dependencies. Users with that approach will need to switch to using `config.npmrc`. Apart from that, everything else works.

**Comment by rarkins** [comment]
We currently use https://github.com/rexxars/registry-auth-token, which in turn calls https://github.com/dominictarr/rc, which employs `fs.statSync` and tries a lot of files.

**Comment by rarkins** [comment]
One problem:
- If you have `~/.npmrc` and `/current/pwd/.npmrc` then they get *merged*
- If you configure `config.npmrc` and there's `/current/pwd/.npmrc` then the configured `npmrc` completely overrides it

Is that approach still valid?

**Comment by rarkins** [comment]
Current behavior:
- config.npmrc overrides all
- `$pwd/.npmrc` takes precedence over and is not merged with `~/.npmrc`
- `~/.npmrc` is the fallback

**Comment by raspy8766** [comment]
> @viceice what are your thoughts on deprecating `npmrc` as a config field? Options:
> 
> * Document that if scoped lookups are required then these should be committed to the repo in a `.npmrc` (doesn't mean auth needs to be committed.. just scope instructions)
> * Allow converting of `packageRules` to `.npmrc` scope instructions via some type of convention. e.g. packageRule follows convention:
> 
> ```json
> {
>   "hostType": "npm",
>   "packagePatterns": "^@somescope\/",
>   "registryUrls": ["https://my.custom.registry/"]
> }
> ```
> 
> converts to:
> 
> ```
> @somescope:registry=https://my.custom.registry/
> ```


@rarkins @viceice we are having an issue with private scoped npm repo's (using artifactory) and this seems like a great way to fix it.

Currently our `.npmrc` looks something like this:
```ini
registry=https://registry.npmjs.org
@scope:registry=https://hcss.jfrog.io/hcss/api/npm/scope/
//org.jfrog.io/org/api/npm/scope/:_password=${ARTIFACTORY_PASSWORD}
//org.jfrog.io/org/api/npm/scope/:username=${ARTIFACTORY_USERNAME}
//org.jfrog.io/org/api/npm/scope/:always-auth=true
```

We would love to adopt renovate, but this is proving to be a serious barrier to entry. We have about 30+ repos using repo defined `.npmrc` files, and I think the only way to get this to work would be to use the `${NPM_TOKEN}` as a workaround. We've also looked at generating an `.npmrc` dynamically with the `hostRules` setting, but without the support for scoped packages we end up getting an `.npmrc` looking something like this instead:

```ini
registry=https://hcss.jfrog.io/hcss/api/npm/scope/
//org.jfrog.io/org/api/npm/scope/:_password=${ARTIFACTORY_PASSWORD}
//org.jfrog.io/org/api/npm/scope/:username=${ARTIFACTORY_USERNAME}
//org.jfrog.io/org/api/npm/scope/:always-auth=true
```

It's also worth noting, Artifactory does not yet support the preferred [`_authToken` authentication](https://www.jfrog.com/jira/browse/RTFACT-19164), just [basic auth](https://www.jfrog.com/confluence/display/JFROG/npm+Registry#npmRegistry-AuthenticatingthenpmClient) which makes using `NPM_TOKEN` alone insufficient without storing the username in plain text.

Lastly, I tried to encrypt the entire `.npmrc` file replacing newlines as specified in the docs, unfortunately it seems to be too large and the [renovate encrypt app](https://app.renovatebot.com/encrypt) just returns `false`

**Comment by raspy8766** [comment]
[Found a solution](https://github.com/renovatebot/renovate/discussions/9801#discussioncomment-697726), providing a "template" `.npmrc` file with the registries predefined seems to do the trick.

**Comment by rarkins** [comment]
@raspy8766 if you have the time to write that solution up in any more detail either here or elsewhere then it would be great for others.

**Comment by raspy8766** [comment]
@rarkins sure, this is basically it:
```json
{
  "npmrc": "registry=https://registry.npmjs.org/\n@scope:registry=https://org.jfrog.io/org/api/npm/scope/",
  "hostRules": [
    {
      "hostType": "npm",
      "baseUrl": "https://org.jfrog.io/org/api/npm/scope/",
      "authType": "Basic",
      "encrypted": {
        "token": "1234-abcd-efghijk"
      }
    }
  ]
}
```

**Comment by raulfdm** [comment]
> @rarkins sure, this is basically it:
> 
> ```json
> {
>   "npmrc": "registry=https://registry.npmjs.org/\n@scope:registry=https://org.jfrog.io/org/api/npm/scope/",
>   "hostRules": [
>     {
>       "hostType": "npm",
>       "baseUrl": "https://org.jfrog.io/org/api/npm/scope/",
>       "authType": "Basic",
>       "encrypted": {
>         "token": "1234-abcd-efghijk"
>       }
>     }
>   ]
> }
> ```

The `encrypted` did the job for me using artifactory.
